client/telemetry/src/lib.rs

```diff
@@ -16,339 +16,486 @@
-			None
+impl TelemetryWorker {
+	pub(crate) fn new(buffer_size: usize, transport: WsTrans) -> Self {
+		let (message_sender, message_receiver) = mpsc::channel(buffer_size);
+		let (register_sender, register_receiver) = mpsc::unbounded();
+
+		Self {
+			message_receiver,
+			message_sender,
+			register_receiver,
+			register_sender,
+			transport,
 		}
-	};
+	}
 
-	Telemetry {
-		inner: Arc::new(Mutex::new(TelemetryInner {
-			worker,
-			receiver,
-		})),
-		_guard: Arc::new(guard),
+	/// Get a new [`TelemetryHandle`].
+	///
+	/// This is used when you want to register a new telemetry for a Substrate node.
+	pub fn handle(&self) -> TelemetryHandle {
+		TelemetryHandle {
+			message_sender: self.register_sender.clone(),
+		}
 	}
-}
 
-/// Event generated when polling the worker.
-#[derive(Debug)]
-pub enum TelemetryEvent {
-	/// We have established a connection to one of the telemetry endpoint, either for the first
-	/// time or after having been disconnected earlier.
-	Connected,
-}
+	/// Get a clone of the channel's `Sender` used to send telemetry events.
+	pub(crate) fn message_sender(&self) -> mpsc::Sender<TelemetryMessage> {
+		self.message_sender.clone()
+	}
 
-impl Stream for Telemetry {
-	type Item = TelemetryEvent;
-
-	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
-		let before = Instant::now();
-
-		// Because the `Telemetry` is cloneable, we need to put the actual fields behind a `Mutex`.
-		// However, the user is only ever supposed to poll from one instance of `Telemetry`, while
-		// the other instances are used only for RAII purposes.
-		// We assume that the user is following this advice and therefore that the `Mutex` is only
-		// ever locked once at a time.
-		let mut inner = match self.inner.try_lock() {
-			Some(l) => l,
-			None => {
-				warn!(
+	/// Run the telemetry worker.
+	///
+	/// This should be run in a background task.
+	pub async fn run(self) {
+		let Self {
+			mut message_receiver,
+			message_sender: _,
+			mut register_receiver,
+			register_sender: _,
+			transport,
+		} = self;
+
+		let mut node_map: HashMap<Id, Vec<(u8, Multiaddr)>> = HashMap::new();
+		let mut node_pool: HashMap<Multiaddr, _> = HashMap::new();
+
+		// initialize the telemetry nodes and register new notifiers
+		async fn process_register(
+			input: Option<Register>,
+			node_pool: &mut HashMap<Multiaddr, Node<WsTrans>>,
+			node_map: &mut HashMap<Id, Vec<(u8, Multiaddr)>>,
+			transport: WsTrans,
+		) {
+			let input = if let Some(x) = input {
+				x
+			} else {
+				log::error!(
 					target: "telemetry",
-					"The telemetry seems to be polled multiple times simultaneously"
+					"Unexpected end of stream. This is a bug.",
 				);
-				// Returning `Pending` here means that we may never get polled again, but this is
-				// ok because we're in a situation where something else is actually currently doing
-				// the polling.
-				return Poll::Pending;
+				return;
+			};
```

Comment:
Can be an expect, as we hold sender/receiver and this can not happen.

Reply:
Fixed in f462418e2